### PR TITLE
cleanup Ordinal

### PR DESCRIPTION
**Pull request by timotheecour**
I'm curious what's the point of having both Ordinal and SomeOrdinal, if anyone can chime in; 

regardless, this PR should be correct and simplifies things.
restarting CI because of https://github.com/nim-lang/Nim/issues/13477 flaky test again...

### CODE REVIEW

File: lib/system/basic_types.nim [hunk 1]
```diff
@@ -20,10 +20,6 @@ const
   off* = false  ## Alias for ``false``.
 
 type
-  Ordinal*[T] {.magic: Ordinal.} ## Generic ordinal type. Includes integer,
-                                 ## bool, character, and enumeration types
-                                 ## as well as their subtypes.
-
```
Comment (krux02) [inline lib/system/basic_types.nim:26]: Why did you move this declaration to `system.nim`?

Comment (timotheecour) [reply lib/system/basic_types.nim:26]: it's an include file so doesn't make a difference; I moved it to fix bootstrap; once we upgrade nim version in csources to 1.0.X, we can put it back there

**Comment by Araq** [comment]
This is so random... How is adding another ``when`` to system.nim a cleanup? I mean, ok, eventually it will result in a cleanup but right now, it doesn't. Here are the numbers:

+36 −10 

After your "cleanup", 36 lines added, 10 removed.

**Comment by krux02** [comment]
Hmm, somehow I thought this PR included a fix for the ``Ordinal`` type magic. But the type is already fixed. It only cleans up `inc` and `dec`. Not sure if that is worth the introduction of a bootstrapping symbol.

@Araq your metric to measure cleanup quality (``+36 −10``) contains the added lines for the test, and the lines that were changed in the other test. I don't that it should.

**Comment by Araq** [comment]
@krux02 Tests also need to be maintained. But regardless of that the ``when`` inside system.nim is a real addition.

**Comment by timotheecour** [comment]
it is indeed a code cleanup, using following definition for code quality: **implementation details should not leak to client code**.

## before PR
things like
```nim
var s = ""
s.inc
```
produce really surprising error msgs for users who expect Ordinal to include uint and uint64:
```
Error: type mismatch: got <string> but expected one of: proc inc[T: Ordinal | uint | uint64](x: var T; y = 1)
```

and then you look up the definitions and you see things like:
```
proc hash*(x: int|int64|uint|uint64|char|Ordinal): Hash {.inline.} =
proc inc*[T: Ordinal|uint|uint64](x: var T, y = 1) {.magic: "Inc", noSideEffect.}
proc dec*[T: Ordinal|uint|uint64](x: var T, y = 1) {.magic: "Dec", noSideEffect.}
```
which reinforces the suspicion about `Ordinal`; in other cases you write code with a signature `Foo[T: Ordinal](a: T)`, which works locally but later fails cryptically in CI because of bootstrap.

Implementation detail leaked to all callers (ie, client code of `Ordinal`), as well as error msgs.

## after PR
the implementation detail is all encapsulated in a single place that's easy to grok, and the "dirty parts" aren't leaked to code using it
```nim
when defined(nimHashOrdinalFixed):
  type Ordinal*[T] {.magic: Ordinal.}
else: 
  type
    OrdinalImpl[T] {.magic: Ordinal.}
    Ordinal* = OrdinalImpl | uint | uint64
```

Btw, this is no different from other existing bootstrap logic (90 occurences of `when defined(nimHasX)`) eg:
```nim
when defined(nimvarargstyped):
  proc echo*(x: varargs[typed, `$`]) {.magic: "Echo", tags: [WriteIOEffect], benign, sideEffect.}
else:
  proc echo*(x: varargs[untyped, `$`]) {.magic: "Echo", tags: [WriteIOEffect], benign, sideEffect.}
```

**Comment by Araq** [comment]
Convincing, thanks for the explanation.
